dev/RadioButtons/RadioButtons.cpp

```diff
@@ -7,136 +7,634 @@
+                            {
+                                if (auto const selectedItem = repeater.TryGetElement(selectedIndex.GetAt(0)))
+                                {
+                                    args.TrySetNewFocusedElement(selectedItem);
+                                }
+                            }
+                            // If focus is coming from the first element to the last element (or last to first), via keyboard we know (suspect)
+                            // that RadioButton has interfered and we want to correct. 
+                            else if (oldFocusedRepeaterIndex == 0 ||
+                                     oldFocusedRepeaterIndex == repeaterItemCount - 1)
+                            {
+                                if (auto const newFocusedElement = args.NewFocusedElement())
+                                {
+                                    if (auto const newFocusedElementAsUIElement = newFocusedElement.try_as<winrt::UIElement>())
+                                    {
+                                        auto const newFocusedRepeaterIndex = repeater.GetElementIndex(newFocusedElementAsUIElement);
+                                        if ((oldFocusedRepeaterIndex == 0 && newFocusedRepeaterIndex == repeaterItemCount - 1) ||
+                                            (oldFocusedRepeaterIndex == repeaterItemCount - 1 && newFocusedRepeaterIndex == 0))
+                                        {
+                                            args.TrySetNewFocusedElement(oldFocusedElement);
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
     }
+}
 
-    UpdateItemsSource();
+void RadioButtons::OnRepeaterLoaded(const winrt::IInspectable&, const winrt::RoutedEventArgs&)
+{
+    if (auto const repeater = m_repeater.get())
+    {
+        if (m_testHooksEnabled)
+        {
+            AttachToLayoutChanged();
+        }
+
+        UpdateSelectedIndex();
+        UpdateSelectedItem();
+        UpdateMaximumColumns();
+        OnRepeaterCollectionChanged(nullptr, nullptr);
+    }
+}
+
+void RadioButtons::KeyDownHandler(const winrt::IInspectable&, const winrt::KeyRoutedEventArgs& args)
+{
+    if (args.Key() == winrt::VirtualKey::Control)
+    {
+        m_isControlDown = true;
+    }
+}
+
+void RadioButtons::KeyUpHandler(const winrt::IInspectable&, const winrt::KeyRoutedEventArgs& args)
+{
+    if (args.Key() == winrt::VirtualKey::Control)
+    {
+        m_isControlDown = false;
+    }
 }
 
-void RadioButtons::OnListViewLoaded(const winrt::IInspectable& /*sender*/, const winrt::RoutedEventArgs& args)
+void RadioButtons::OnChildKeyDown(const winrt::IInspectable&, const winrt::KeyRoutedEventArgs& args)
 {
-    if (auto listView = m_listView.get())
+    if (args.OriginalKey() == winrt::VirtualKey::Down)
```

Comment:
you should check if args.Handled is already true and not do any of this if that is the case. Also, a switch case here might make it more readable.

Reply:
I thought that the only way to receive events which are already handled is to register via the AddHandler method with the ForHandled flag set to true. I haven't done that so I don't think it is possible for me to see a handled event. 

---
In reply to: [347505129](https://github.com/microsoft/microsoft-ui-xaml/pull/1623#discussion_r347505129) [](ancestors = 347505129)